starter-code/src/memory.js

```diff
@@ -1,15 +1,32 @@
-// var MemoryGame = function (cards) {
-//   this.cards = cards;
-// };
+var MemoryGame = function(cards) {
+  this.cards = cards;
+  this.pickedCards = [];
+  this.pairsClicked = 0;
+  this.pairsGuessed = 0;
+};
 
-// MemoryGame.prototype.shuffleCard = function (cardsArr) {
+MemoryGame.prototype.shuffleCard = function(cardsArr) {
+  for (let i = cardsArr.length - 1; i > 0; i--) {
+    const j = Math.floor(Math.random() * (i + 1));
+    [cardsArr[i], cardsArr[j]] = [cardsArr[j], cardsArr[i]];
+  }
+  return cardsArr;
+};
 
-// };
+MemoryGame.prototype.checkIfPair = function(firstCard, secondCard) {
+  this.pairsClicked += 1;
+  if (firstCard === secondCard) {
+    this.pairsGuessed += 1;
+    return true;
+  } else {
+    return false;
+  }
+};
 
-// MemoryGame.prototype.checkIfPair = function (firstCard, secondCard) {
-
-// }
-
-// MemoryGame.prototype.finished = function () {
-
-// };
+MemoryGame.prototype.finished = function() {
+  if (this.pairsGuessed === 12) {
```

Comment:
This works for this particular set of cards, but notice that the constructor receives a "cards" argument, so if we wanted to start a game with a different number of cards, it wouldn't be 12 turns. To solve this, you can simply do this:
`
if ( this.pairsGuessed === this.cards / 2 ) {
...
...
...
}
`